Hide masterbar when Calypso is viewed from WC apps

guarani opened the pull request:
#### Changes proposed in this Pull Request

* Add a new utility method, `isWcMobileApp`, to detect when Calypso is being displayed from within the WooCommerce apps (WCiOS or WCAndroid).
* Hide the `masterbar` if Calypso is accessed from the WooCommerce app (WPiOS or WCAndroid).

| Before | After |
| - | - |
| <img width="250" alt="Before: the masterbar is shown" src="https://user-images.githubusercontent.com/1898325/124011602-31ebb580-d9ae-11eb-9e68-6a2ef0ae21cd.png"> | <img width="250" alt="After: the masterbar is hidden" src="https://user-images.githubusercontent.com/1898325/124011554-23050300-d9ae-11eb-9fe7-dc4c4d6c3412.png"> |


#### Testing instructions

* Checkout this branch and run the app locally: `yarn && yarn start`
* Open http://calypso.localhost:3000
* Verify the masterbar is visible as usual
* Change the [user agent to that of WCiOS](https://github.com/woocommerce/woocommerce-ios/blob/4fe952211777fad134084cbc19d4db229a37c8c9/Networking/Networking/Settings/UserAgent.swift#L40): `Mozilla/5.0 (iPhone; CPU iPhone OS 12_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/16B91 wc-ios/12.1`
* Reload the page and verify that the masterbar is hidden
* Change the user agent to that of WCAndroid: `Mozilla/5.0 (iPhone; CPU iPhone OS 12_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/16B91 wc-android/12.1`
* Reload the page and verify that the masterbar is still hidden

Tip: To change the user agent in Chrome, see below screenshots (taken from https://github.com/Automattic/wp-calypso/pull/31830):

<img width="404" alt="55129228-ab562700-5159-11e9-9a75-d05cfe97b65e" src="https://user-images.githubusercontent.com/1898325/124031455-556e2a80-d9c5-11eb-9a74-02a5315f4bf9.png">
<img width="900" alt="55129288-df314c80-5159-11e9-8278-f14471c9eb2f" src="https://user-images.githubusercontent.com/1898325/124031451-54d59400-d9c5-11eb-9924-d252d408879f.png">

Fixes #53982

Commit:
Hide masterbar when adding credit card via WC apps

The WooCommerce apps show Calypso's "add credit card" pages, but have no need for the masterbar. This PR uses the browser's (or webview's) user agent to hide the masterbar when the WCiOS or WCAndroid user agent is present.

guarani commented on client/me/purchases/add-new-payment-method/index.jsx:
```diff
@@ -56,6 +58,10 @@ function AddNewPaymentMethod() {
 	const title = `${ titles.activeUpgrades } › ${ addPaymentMethodTitle }`;
 
+	if ( isWcMobileApp() ) {
+		reduxDispatch( hideMasterbar() );
```
I just need to figure out if we need to explicitly call `showMasterbar` at some point, to make the `masterbar` visible again on other pages. Not sure yet if it's needed, though.

guarani replied:
@hichamboushaba on second thought, should we be hiding the `masterbar` on **_all_** pages, similar to https://github.com/Automattic/wp-calypso/pull/54189, or is the goal here to only hide it on `/me/purchases/add-credit-card` and nowhere else?

hichamboushaba replied:
@guarani our goal was to hide everywhere in case we need to show the web UI elsewhere, though the immediate need is only for `/me/purchases/add-credit-card`.

guarani replied:
Sounds good! I think it simplifies the implementation so I made the change to hide the `masterbar` on all pages when viewed from the WC apps.

guarani replied:
Making this as resolved.

Commit:
Fix inline docs for isWcMobileApp

Commit:
Hide masterbar always in WooCommerce apps

This makes the masterbar hidden on any page, not just on a specific page as was the case before.

Commit:
Clean up unused code

hichamboushaba commented on client/lib/mobile-app/index.js:
```diff
@@ -9,3 +9,15 @@ export function isWpMobileApp() {
+	if ( typeof navigator === 'undefined' ) {
+		return false;
+	}
+	return navigator.userAgent && /wc-(android|iphone)/.test( navigator.userAgent );
```
Can you please update the useragent of the WCiOS? it is `wc-ios` (source: https://github.com/woocommerce/woocommerce-ios/blob/4fe952211777fad134084cbc19d4db229a37c8c9/Networking/Networking/Settings/UserAgent.swift#L40)

guarani replied:
Corrected, thanks!

Commit:
Fixed useragent detection for WooCommerce apps

Commit:
Add isWcMobileApp tests

Run using `yarn run test-client client/lib/mobile-app/test/index.js`

nsakaimbo reviewed (approved):
Tested and confirmed it does what it says on the tin! Thanks again for the detailed instructions - code changes look good to me. 👍

guarani commented:
Thanks for much for the review, @nsakaimbo!

sirbrillig reviewed (commented):
I haven't tested this but no objections from payments. 👍

guarani commented:
> I haven't tested this but no objections from payments. 👍

Thank you @sirbrillig for taking a look!

I just saw this [release checklist](https://github.com/Automattic/wp-calypso/blob/trunk/docs/merge-checklist.md) which I hadn't seen earlier, so I'll go through it. 
One question I have is: Do I merge this myself or does someone else need to approve or push the merge button themself?

michaeldcain commented:
@guarani, if you're comfortable, you can follow the instructions to merge/deploy Calypso here: PCYsg-5XR-p2

If you have any questions, you can ask in `#calypso` in Slack.

guarani commented:
Thank you @michaeldcain!

Commit:
Merge branch 'trunk' into update/hide-masterbar-in-woocommerce-apps

noahtallen commented:
Hi! I wanted to mention that you don't need to worry about the comments with`This PR modifies the release build for...`. 

I've just merged a PR which deletes stale comments which no longer apply to the PR, so if you rebase on trunk again, those should be deleted. I'm also about ready to [merge a fix](https://github.com/Automattic/wp-calypso/pull/54809) which will stop those messages from showing up altogether in situations like this PR :)

guarani commented:
Thanks for much for commenting @noahtallen, I actually had noted those down to ask about, so you've answered my questions already! 🙇‍♂️

Commit:
Merge branch 'trunk' into update/hide-masterbar-in-woocommerce-apps

guarani commented:
I went through the [merge checklist](https://github.com/Automattic/wp-calypso/blob/trunk/docs/merge-checklist.md) and found a few potential issues/unknows which I've highlighted with a ⚠️ icon:

> - Run `yarn test` from the calypso root to make sure all tests pass.

⚠️ I get some failing tests when running `yarn test` on this branch, but the tests also fail on `trunk`.

```
● Webpack RTL Plugin › Bundling › should create a second bundle
● Webpack RTL Plugin › Bundling › should contain the correct content
```

> 2. Test in IE11+, Chrome, Firefox, Mac/Win on the desktop.

⚠️ Can't test on IE11+ due to lack of Win machine. Tested on Chrome, Firefox and Safari on Mac.

noahtallen commented:
@guarani You should be ok for both of those items. For the tests, all the important ones are run as part of the GH checks, which are passing. We are also in the process of phasing out IE11 support, so don't worry about that either :)

guarani commented:
Sounds good, @noahtallen! I forgot to mention how those tests only failed locally, so it didn't seem like a real issue. Also, FWIW, I ran the failing test in isolation (`yarn run test-packages packages/webpack-rtl-plugin/test/index.js`) and it passed consistently.
